packages/repository/test/unit/decorator/relation.decorator.unit.ts

```diff
@@ -0,0 +1,120 @@
+import {expect} from '@loopback/testlab';
+import {Entity} from '../../../src/model';
+import {hasMany, RELATIONS_KEY, RelationType, property} from '../../../src';
+import {MetadataInspector} from '@loopback/context';
+
+describe('relation decorator', () => {
+  context('hasMany', () => {
+    it('infers property type from property decorator', () => {
+      class Address extends Entity {
+        addressId: number;
+        street: string;
+        province: string;
+      }
+
+      class AddressBook extends Entity {
+        id: number;
+
+        @hasMany()
+        @property.array(Address)
```

Comment:
Considering that `@property.array() @hasMany` does not work (see the test below), and `@hasMany` accepts `modelTo`, is it really necessary to support this usage of our decorators? To me, this looks like unnecessary long way how to express a model relation. What if we supported only the following form?

```ts
class AddressBook extends Entity {
  // ...

  @hasMany({modelTo: Address})
  addresses: Address[],
}
```

Alternatively, we can move `modelTo` from options property into a decorator argument:

```ts
@hasMany(Address)
```

Thoughts?

Reply:
We've decided to get rid of `modelTo` key as it is not needed for the constrained repository resolution.

Reply:
IIUC, one is going to define a has-many relation as follows:

```ts
@hasMany()
@property.array(Address)
addresses: Address[];
```

I am arguing that such usage is unnecessary long and makes room for mistakes, e.g. the developer uses `@hasMany` but forgets `@property`. IMO, my proposed version would be more practical:

```ts
@hasMany(Address)
addresses: Address[]
```

Under the hood, `@hasMany` can call `@property` for the developer. I am imagining something along the following lines (the type names may be wrong):

```ts
function hasMany<T>(targetModel: Class<T>, relationConfig?: HasManyMetadata) {
  return function(target: Object, key: string) {
    property.array(targetClass)(target, key);
    // continue with hasMany-related code as in your proposal
  };
}
```
  
Let's discuss! Are there any arguments in favor of your proposal and/or against my proposal?

Reply:
FWIW, in Java world, relation properties seem to be decorated with the relational decorator only too.

See e.g. https://stackoverflow.com/questions/37362676/spring-boot-onetomany-with-jpa#37371680

```java
@Entity
@Table(name="Company_new")
public class Company {
//...
@Column(name="name")
private String name;

@OneToMany(mappedBy="company",cascade=CascadeType.ALL)
private Set<User> users;
}
```

Reply:
I think it's fine to merge `@hasMany` and `@property` together, but if we do we should make it clear in our docs that properties using `@relation` decorators should not need to use `@property`. A concern I have with this approach is the fact that we'd have to support a number of parameters for the decorator, with examples being `PropertyDefinition`, `HasManyMetadata` and any other form of sugar. I fear that the decorator can get overloaded very fast this way.

Not opposed to the idea though as long as the parameters that the decorator can take doesn't look ugly. Maybe something like this:
```ts
@hasMany(Address // where this is sugar for PropertyDefinition
, {
  fk: foreignKey
})
```

Reply:
>  A concern I have with this approach is the fact that we'd have to support a number of parameters for the decorator, with examples being PropertyDefinition, HasManyMetadata and any other form of sugar. I fear that the decorator can get overloaded very fast this way.

Hmm, this is a good point that I haven't considered.

On the other hand, what kind of PropertyDefinition metadata do you expect to see in relational properties? Wouldn't most of that metadata be defined by the framework via the relation?